src/main/java/htsjdk/samtools/cram/compression/rans/ransnx16/RANSNx16Decode.java

```diff
@@ -0,0 +1,488 @@
+        }
+        else {
+            freqTableSource = inBuffer;
+        }
+        readFrequencyTableOrder1(freqTableSource, shift);
+        final ArithmeticDecoder[] D = getD();
+        final RANSDecodingSymbol[][] syms = getDecodingSymbols();
+        final int outputSize = outBuffer.remaining();
+        inBuffer.order(ByteOrder.LITTLE_ENDIAN);
+
+        // Nway parallel rans states. Nway = 4 or 32
+        final int Nway = ransNx16Params.getNumInterleavedRANSStates();
+        final long[] rans = new long[Nway];
+        final int[] interleaveStreamIndex = new int[Nway];
+        final int[] context = new int[Nway];
+        final int[] symbol = new int[Nway];
+
+        // size of interleaved stream = outputSize / Nway
+        // For Nway = 4, division by 4 is the same as right shift by 2 bits
+        // For Nway = 32, division by 32 is the same as right shift by 5 bits
+        final int interleaveSize = (Nway==4) ? (outputSize >> 2): (outputSize >> 5);
+
+        for (int r=0; r<Nway; r++){
+
+            // initialize rans, interleaveStreamIndex and context arrays
+            rans[r] = inBuffer.getInt();
+            interleaveStreamIndex[r] = r * interleaveSize;
+            context[r] = 0;
+        }
+
+        while (interleaveStreamIndex[0] < interleaveSize) {
+            for (int r = 0; r < Nway; r++){
+                symbol[r] = 0xFF & D[context[r]].reverseLookup[Utils.RANSGetCumulativeFrequency(rans[r], shift)];
+                outBuffer.put(interleaveStreamIndex[r],(byte) symbol[r]);
+                rans[r] = syms[context[r]][symbol[r]].advanceSymbolStep(rans[r], shift);
+                rans[r] = Utils.RANSDecodeRenormalizeNx16(rans[r], inBuffer);
+                context[r]=symbol[r];
+            }
+            for (int r = 0; r < Nway; r++){
+                interleaveStreamIndex[r]++;
+            }
+        }
+
+        // Deal with the remaining elements
+        for (; interleaveStreamIndex[Nway - 1] < outputSize; interleaveStreamIndex[Nway - 1]++) {
+            symbol[Nway - 1] = 0xFF & D[context[Nway - 1]].reverseLookup[Utils.RANSGetCumulativeFrequency(rans[Nway - 1], shift)];
+            outBuffer.put(interleaveStreamIndex[Nway - 1], (byte) symbol[Nway - 1]);
+            rans[Nway - 1] = syms[context[Nway - 1]][symbol[Nway - 1]].advanceSymbolNx16(rans[Nway - 1], inBuffer, shift);
+            context[Nway - 1] = symbol[Nway - 1];
+        }
+        return outBuffer;
+    }
+
+    private void readFrequencyTableOrder0(
+            final ByteBuffer cp) {
+
+        // Use the Frequency table to set the values of Frequencies, Cumulative Frequency
+        // and Reverse Lookup table
+
+        final ArithmeticDecoder decoder = getD()[0];
+        final RANSDecodingSymbol[] decodingSymbols = getDecodingSymbols()[0];
+        final int[] alphabet = readAlphabet(cp);
+        int cumulativeFrequency = 0;
+        final int[] frequencies = new int[Constants.NUMBER_OF_SYMBOLS];
```

Comment:
I wouldn't necessarily change this now, since there are lots of other issues we need to address, but this intermediate array is pretty much throw-away. I think we could be using the `decoder` and `decodingSymbols` arrays directly, and eliminate the need for these intermediate allocations entirely. Again, lets not change this now, but it something to keep in mind for future code where we need to be mindful of allocation/performance.

Reply:
I made a few quick changes. We now use decoder instead of a new frequencies array. I have also removed cumulativeFreq from ArithmeticDecoder.